File: app/src/main/java/io/github/ovso/psytest/ui/main/fragment/VideoFragmentPresenterImpl.java
package io.github.ovso.psytest.ui.main.fragment;

import android.content.ActivityNotFoundException;
import android.os.Bundle;
import android.text.TextUtils;
import io.github.ovso.psytest.R;
import io.github.ovso.psytest.data.KeyName;
import io.github.ovso.psytest.data.network.SearchRequest;
import io.github.ovso.psytest.data.network.model.SearchItem;
import io.github.ovso.psytest.ui.main.fragment.adapter.VideoAdapterDataModel;
import io.github.ovso.psytest.utils.ResourceProvider;
import io.github.ovso.psytest.utils.SchedulersFacade;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import java.util.Collections;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Getter;
import timber.log.Timber;

public class VideoFragmentPresenterImpl implements VideoFragmentPresenter {

  private CompositeDisposable compositeDisposable = new CompositeDisposable();
  private VideoFragmentPresenter.View view;
  private SearchRequest searchRequest;
  private ResourceProvider resourceProvider;
  private SchedulersFacade schedulersFacade;
  private VideoAdapterDataModel<SearchItem> adapterDataModel;
  private String nextPageToken;
  private String q;
  private int position;

  public VideoFragmentPresenterImpl(VideoFragmentPresenter.View $view,
      SearchRequest $searchRequest, ResourceProvider $resourceProvider,
      SchedulersFacade $schedulersFacade, VideoAdapterDataModel<SearchItem> $adapterDataModel) {
    view = $view;
    searchRequest = $searchRequest;
    resourceProvider = $resourceProvider;
    schedulersFacade = $schedulersFacade;
    adapterDataModel = $adapterDataModel;
  }

  @Override public void onActivityCreated(Bundle args) {
    Timber.d("onActivityCreated");
    view.setupRecyclerView();
    view.setupSwipeRefresh();
    view.showLoading();
    view.setupAdListener();
    position = args.getInt(KeyName.POSITION.get());
    q = resourceProvider.getStringArray(R.array.q)[position];
    Disposable disposable = searchRequest.getResult(q, nextPageToken)
        .subscribeOn(schedulersFacade.io())
        .observeOn(schedulersFacade.ui())
        .subscribe(
            search -> {
              nextPageToken = search.getNextPageToken();
              List<SearchItem> items = search.getItems();
              shuffle(items);
              adapterDataModel.addAll(items);
              view.refresh();
              view.setLoaded();
              view.hideLoading();
            }, throwable -> {
              Timber.d(throwable);
              view.hideLoading();
            });
    compositeDisposable.add(disposable);
  }

  @Override public void onDestroyView() {
    compositeDisposable.clear();
  }

  @Override public void onItemClick(SearchItem data) {
    try {
      view.showVideo(data.getId().getVideoId());
    } catch (ActivityNotFoundException e) {
      e.printStackTrace();
      view.showYoutubeUseWarningDialog();
    }
  }

  @Override public void onLoadMore() {
    if (!TextUtils.isEmpty(nextPageToken) && !TextUtils.isEmpty(q)) {
      Disposable disposable = searchRequest.getResult(q, nextPageToken)
          .subscribeOn(schedulersFacade.io())
          .observeOn(schedulersFacade.ui())
          .subscribe(
              search -> {
                nextPageToken = search.getNextPageToken();
                List<SearchItem> items = search.getItems();
                shuffle(items);
                adapterDataModel.addAll(items);
                view.refresh();
                view.setLoaded();
              }, Timber::d);
      compositeDisposable.add(disposable);
    }
  }

  @Override public void onRefresh() {
    adapterDataModel.clear();
    view.refresh();
    nextPageToken = null;
    q = resourceProvider.getStringArray(R.array.q)[position];
    Disposable disposable = searchRequest.getResult(q, nextPageToken)
        .subscribeOn(schedulersFacade.io())
        .observeOn(schedulersFacade.ui())
        .subscribe(
            search -> {
              nextPageToken = search.getNextPageToken();
              List<SearchItem> items = search.getItems();
              shuffle(items);
              adapterDataModel.addAll(items);
              view.refresh();
              view.setLoaded();
              view.hideLoading();
            }, throwable -> {
              Timber.d(throwable);
              view.hideLoading();
            });
    compositeDisposable.add(disposable);
  }

  @Override public boolean onOptionsItemSelected(int itemId) {
    String url = Portal.toUrl(itemId, q);
    view.navigateToWeb(url);
    return true;
  }

  private <E> void shuffle(List<E> $items) {
    Collections.shuffle($items);
  }

  @Getter @AllArgsConstructor enum Portal {
    GOOGLE(R.id.action_google, "https://google.co.kr/search?q="),
    NAVER(R.id.action_naver, "https://m.search.naver.com/search.naver?where=m_video&query="),
    DAUM(R.id.action_daum, "https://m.search.daum.net/search?w=vclip&q=");

    private int id;
    private String url;

    public static Portal toType(int id) {
      for (Portal portal : Portal.values()) {
        if (portal.id == id) {
          return portal;
        }
      }

      return GOOGLE;
    }

    public static String toUrl(int act_id, String q) {
      Portal portal = toType(act_id);
      switch (portal) {

        case GOOGLE:
          return Portal.GOOGLE.getUrl() + q + "&tbm=vid";
        case NAVER:
          return Portal.NAVER.getUrl() + q;
        case DAUM:
          return Portal.DAUM.getUrl() + q;
        default:
          return Portal.GOOGLE.getUrl() + q + "&tbm=vid";
      }
    }

  }
}

File: app/src/main/java/io/github/ovso/psytest/ui/video/VideoFragmentPresenterImpl.kt
package io.github.ovso.psytest.ui.video

import android.content.ActivityNotFoundException
import android.os.Bundle
import io.github.ovso.psytest.R
import io.github.ovso.psytest.Security
import io.github.ovso.psytest.data.KeyName
import io.github.ovso.psytest.data.network.SearchRequest
import io.github.ovso.psytest.data.network.model.Search
import io.github.ovso.psytest.data.network.model.SearchItem
import io.github.ovso.psytest.exts.plusAssign
import io.github.ovso.psytest.ui.video.VideoFragmentPresenter.View
import io.github.ovso.psytest.ui.video.adapter.VideoAdapterDataModel
import io.github.ovso.psytest.utils.ResourceProvider
import io.github.ovso.psytest.utils.SchedulersFacade
import io.reactivex.SingleObserver
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import timber.log.Timber
import java.util.Collections.shuffle

class VideoFragmentPresenterImpl(
  private val view: View,
  private val searchRequest: SearchRequest,
  private val resourceProvider: ResourceProvider,
  private val schedulersFacade: SchedulersFacade,
  private val adapterDataModel: VideoAdapterDataModel<SearchItem>
) : VideoFragmentPresenter {

  private val compositeDisposable = CompositeDisposable()
  private var nextPageToken: String? = null
  private var q: String? = null
  private var position: Int = 0

  private fun getParams(): Map<String, Any> = mapOf(
      KeyName.KEY.get() to Security.KEY.value,
      KeyName.Q.get() to (q ?: resourceProvider.getString(R.string.app_name)),
      KeyName.MAX_RESULTS.get() to 3,
      KeyName.ORDER.get() to "viewCount",
      KeyName.TYPE.get() to "video",
      KeyName.VIDEO_SYNDICATED.get() to "any",
      KeyName.PART.get() to "snippet"
  ).apply {
    if (nextPageToken.isNullOrEmpty().not()) {
      KeyName.PAGE_TOKEN.get() to nextPageToken
    }
  }

  override fun onActivityCreated(args: Bundle) {
    view.setupRecyclerView()
    view.setupSwipeRefresh()
    view.showLoading()
    view.setupAdListener()
    view.addEvent()
    view.showTitle(args.getString("title"))
    position = args.getInt(KeyName.POSITION.get())
    q = args.getString("query")

    reqSearch()
  }

  private fun reqSearch() {
    fun shuffle(search: Search): Search {
      shuffle(search.items)
      return search
    }

    compositeDisposable += searchRequest.getResult2(getParams())
        .map(::shuffle)
        .subscribeOn(schedulersFacade.io())
        .observeOn(schedulersFacade.ui())
        .doOnSubscribe { view.showLoading() }
        .doOnSuccess { view.hideLoading() }
        .doOnError { view.hideLoading() }
        .subscribe(::onSuccess, ::onError)
  }

  override fun onDestroyView() {
    compositeDisposable.clear()
  }

  override fun onItemClick(data: SearchItem) {
    try {
      view.showVideo(data.id?.videoId)
    } catch (e: ActivityNotFoundException) {
      e.printStackTrace()
      view.showYoutubeUseWarningDialog()
    }

  }

  private fun onSuccess(search: Search) {
    nextPageToken = search.nextPageToken
    val items = search.items
    adapterDataModel.addAll(items)
    view.refresh()
    view.setLoaded()
  }

  private fun onError(t: Throwable) {
    Timber.e(t)
  }

  override fun onLoadMore() {
    reqSearch()
  }

  @Deprecated("Unsupported")
  override fun onRefresh() {
    adapterDataModel.clear()
    view.refresh()
    nextPageToken = null
    q = resourceProvider.getStringArray(R.array.queries)[position]
    searchRequest.getResult(q, nextPageToken)
        .subscribeOn(schedulersFacade.io())
        .observeOn(schedulersFacade.ui())
        .subscribe(object : SingleObserver<Search> {
          override fun onSubscribe(d: Disposable) {
            compositeDisposable.add(d)
          }

          override fun onSuccess(search: Search) {
            nextPageToken = search.nextPageToken
            val items = search.items
            shuffle(items)
            adapterDataModel.addAll(items)
            view.refresh()
            view.setLoaded()
            view.hideLoading()
          }

          override fun onError(e: Throwable) {
            Timber.e(e)
            view.hideLoading()
          }
        })
  }

  override fun onOptionsItemSelected(itemId: Int): Boolean {
    val url =
      Portal.toUrl(
          itemId, q
      )
    val title = Portal.toType(
        itemId
    )
        .toString()
    view.navigateToWeb(url, title)
    return true
  }

  internal enum class Portal(
    private var id: Int,
    private var url: String
  ) {
    GOOGLE(R.id.action_google, "https://google.co.kr/search?q="),
    NAVER(R.id.action_naver, "https://m.search.naver.com/search.naver?where=m_video&query="),
    DAUM(R.id.action_daum, "https://m.search.daum.net/search?w=vclip&q=");

    companion object {

      fun toType(id: Int): Portal {
        for (portal in values()) {
          if (portal.id == id) {
            return portal
          }
        }

        return GOOGLE
      }

      fun toUrl(
        act_id: Int,
        q: String?
      ): String {
        val portal =
          toType(
              act_id
          )
        return when (portal) {
          GOOGLE -> GOOGLE.url + q!! + "&tbm=vid"
          NAVER -> NAVER.url + q!!
          DAUM -> DAUM.url + q!!
        }
      }
    }
  }
}